xdgopenproxy/xdgopenproxy.go

```diff
@@ -27,37 +27,95 @@ import (
 	"github.com/godbus/dbus"
 )
 
-func Run(urlOrFile string) error {
-	bus, err := dbus.SessionBus()
+type bus interface {
+	Object(name string, objectPath dbus.ObjectPath) dbus.BusObject
+	Close() error
+}
+
+var sessionBus = func() (bus, error) { return dbus.SessionBus() }
+
+type desktopLauncher interface {
+	openFile(path string) error
+	openURL(url string) error
+}
+
+type portalLauncher struct {
+	service dbus.BusObject
+}
+
+func (p *portalLauncher) openFile(filename string) error {
+	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
 	if err != nil {
 		return err
 	}
-	defer bus.Close()
-	launcher := bus.Object("io.snapcraft.Launcher", "/io/snapcraft/Launcher")
-	return launch(launcher, urlOrFile)
+	defer syscall.Close(fd)
+
+	return p.service.Call("org.freedesktop.portal.OpenURI.OpenFile", 0, "", dbus.UnixFD(fd),
+		map[string]dbus.Variant{}).Err
 }
 
-func launch(launcher dbus.BusObject, urlOrFile string) error {
-	if u, err := url.Parse(urlOrFile); err == nil {
-		if u.Scheme == "file" {
-			return openFile(launcher, u.Path)
-		} else if u.Scheme != "" {
-			return openUrl(launcher, urlOrFile)
-		}
+func (p *portalLauncher) openURL(path string) error {
+	return p.service.Call("org.freedesktop.portal.OpenURI.OpenURI", 0, "", path,
+		map[string]dbus.Variant{}).Err
+}
+
+func newPortalLauncher(bus bus) (desktopLauncher, error) {
+	obj := bus.Object("org.freedesktop.portal.Desktop", "/org/freedesktop/portal/desktop")
+
+	err := obj.Call("org.freedesktop.DBus.Peer.Ping", 0).Err
+	if err != nil {
+		return nil, err
```

Comment:
Do we need to make the Ping call?  We could save a message by simply trying the portal API, and falling back to `io.snapcraft.Launcher` if we get an error?  That would reduce the number of method calls in the portals case, and be the same number of calls in the no portals case.

Reply:
Took the advice and I'm refactoring this bit. On top of this, the `org.freedesktop.DBus.Peer.Ping` is blocked by AppArmor (unrelated question whether we should allow `org.freedesktop.Peer` methods? cc @jdstrand ).

I'm slightly concerned about properly handling the error. Right now, the code falls back to next launcher if we get `org.freedesktop.DBus.Error.NoSuchObject`.

Reply:
I think we'd also want to check for NameHasNoOwner for the case where xdg-desktop-portal isn't installed.